src/DotLiquid/StandardFilters.cs

```diff
@@ -667,10 +667,29 @@ private static object DoMathsOperation(object input, object operand, Func<Expres
             if (input == null || operand == null)
                 return null;
 
+            // NOTE(David Burg): Try for maximal precision if the input and operand fit the decimal's range.
+            // This avoids rounding errors in financial arithmetics.
+            // E.g.: 0.1 | Plus 10 | Minus 10 to remain 0.1, not 0.0999999999999996
+            // Otherwise revert to maximum range (possible precision loss).
             if (IsReal(input) || IsReal(operand))
             {
-                input = Convert.ToDouble(input);
-                operand = Convert.ToDouble(operand);
+                try
+                {
+                    input = Convert.ToDecimal(input);
+                    operand = Convert.ToDecimal(operand);
+
+                    return ExpressionUtility
+                        .CreateExpression(
+                            body: operation,
+                            leftType: input.GetType(),
+                            rightType: operand.GetType())
+                        .DynamicInvoke(input, operand);
+                }
+                catch (Exception ex) when (ex is OverflowException || (ex is TargetInvocationException && ex?.InnerException is OverflowException))

```

Comment:
You’re catching OverflowExceptions caused by using Expression.*Checked here, but presumably double operations can now overflow below? I don’t think we should change that behaviour.

Reply:
Fair point. I'll see if I can both get the checked' overflow to be aware when Decimal range is exceeded while not changing the no-overflow current behavior for operations which go even beyond the range of float.

Reply:
Fair point. I'll see if I can both get the checked' overflow to be aware when Decimal range is exceeded while not changing the no-overflow current behavior for operations which go even beyond the range of float.

Reply:
Turns out double is not throwing overflow exception for checked operation, instead it gives infinity and negative infinity. Still researching and testing.

Reply:
That's right, no exception for double arithmetic anyway, as designed and documented: https://docs.microsoft.com/en-us/dotnet/api/system.double?view=netframework-4.8#Exceptions
I will be pushing unit tests demonstrating this anyway.

Reply:
There is a regression however for float and double division by zero. Decimal will throw for those while float and double will return either infinite values or NaN (not a number). I need to catch also the division by zero catching to fallback to double type just like the overflow so the existing float and double behavior is preserved.